src/varity/vcf_to_hgvs/protein.clj

```diff
@@ -211,47 +185,86 @@
+        apply-offset* (fn [exon-ranges*]
+                        (ffirst (alt-exon-ranges exon-ranges* pos ref alt)))
+        pos** (cond (and (= pos* cds-start) (<= pos-start cds-start pos-end))
+                    (get-alt-cds-start-pos cds-start pos-start pos-end exon-ranges pos*)
+                    (and (= pos* cds-end) (<= pos-start cds-end pos-end))
+                    (get-alt-cds-end-pos cds-end pos-start pos-end exon-ranges pos*)
+                    :else
+                    pos*)]
+    (apply-offset* [[pos** pos**]])))
 
 (defn- read-sequence-info
   [seq-rdr rg pos ref alt]
   (if-let [overlap-exon-intron-boundary (overlap-exon-intron-boundary? (:exon-ranges rg) pos ref alt)]
     {:overlap-exon-intron-boundary overlap-exon-intron-boundary}
     (let [{:keys [chr tx-start tx-end cds-start cds-end exon-ranges strand]} rg
-          ref-seq (cseq/read-sequence seq-rdr {:chr chr, :start cds-start, :end cds-end})
-          alt-seq (common/alt-sequence ref-seq cds-start pos ref alt)
-          alt-exon-ranges* (alt-exon-ranges exon-ranges pos ref alt)
-          ref-exon-seq (exon-sequence ref-seq cds-start exon-ranges)
-          ref-up-exon-seq (read-exon-sequence seq-rdr chr tx-start (dec cds-start) exon-ranges)
-          alt-up-exon-seq (make-alt-up-exon-seq ref-up-exon-seq cds-start pos ref alt)
-          ref-down-exon-seq (read-exon-sequence seq-rdr chr (inc cds-end) tx-end exon-ranges)
-          alt-down-exon-seq (make-alt-down-exon-seq ref-down-exon-seq cds-end pos ref alt)
-          alt-exon-seq (exon-sequence alt-seq cds-start alt-exon-ranges*)
-          ter-site-adjusted-alt-seq (make-ter-site-adjusted-alt-seq alt-exon-seq alt-up-exon-seq alt-down-exon-seq
-                                                                    strand cds-start cds-end pos ref)
+          ref-seq (cseq/read-sequence seq-rdr {:chr chr, :start tx-start, :end tx-end})
+          ref-cds-seq (cseq/read-sequence seq-rdr {:chr chr, :start cds-start, :end cds-end})
+          ref-cds-exon-seq (exon-sequence ref-cds-seq cds-start cds-end exon-ranges)
           ref-include-utr-ini-site-boundary (include-utr-ini-site-boundary? rg pos ref alt)
           ref-include-ter-site (include-ter-site? rg pos ref alt)
-          apply-offset* (partial apply-offset pos ref alt alt-exon-ranges* ref-include-ter-site)]
-      {:ref-exon-seq ref-exon-seq
-       :ref-prot-seq (codon/amino-acid-sequence (cond-> ref-exon-seq
+          alt-seq (common/alt-sequence ref-seq tx-start pos ref alt)
+          alt-exon-ranges* (alt-exon-ranges exon-ranges pos ref alt)
+          apply-offset* (partial apply-offset pos ref alt cds-start cds-end exon-ranges)
+          alt-cds-start (apply-offset* cds-start)
+          alt-cds-end (apply-offset* cds-end)
+          alt-tx-end (apply-offset* tx-end)
+          [alt-up-exon-seq alt-cds-exon-seq alt-down-exon-seq] (when (and alt-cds-start alt-cds-end)
```

Comment:
To treat alt upstream and downstream seq more precisely, I fixed alt seq create process.